cpp/src/arrow/compute/kernels/vector_is_monotonic.cc

```diff
@@ -0,0 +1,317 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/aggregate_internal.h"
+#include "arrow/compute/registry.h"
+#include "arrow/util/optional.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+namespace {
+// ----------------------------------------------------------------------
+// IsMonotonic implementation
+
+using IsMonotonicState = OptionsWrapper<IsMonotonicOptions>;
+
+Status IsMonotonicOutput(bool increasing, bool strictly_increasing, bool decreasing,
+                         bool strictly_decreasing, Datum* out) {
+  ARROW_ASSIGN_OR_RAISE(
+      *out, StructScalar::Make({std::make_shared<BooleanScalar>(increasing),
+                                std::make_shared<BooleanScalar>(strictly_increasing),
+                                std::make_shared<BooleanScalar>(decreasing),
+                                std::make_shared<BooleanScalar>(strictly_decreasing)},
+                               {"increasing", "strictly_increasing", "decreasing",
+                                "strictly_decreasing"}));
+  return Status::OK();
+}
+
+template <typename DataType>
+enable_if_floating_point<DataType> IsMonotonicCheck(
+    const typename DataType::c_type& current, const typename DataType::c_type& next,
+    bool* increasing, bool* strictly_increasing, bool* decreasing,
+    bool* strictly_decreasing, const IsMonotonicOptions& options) {
+  // Short circuit for NaNs.
+  // https://en.wikipedia.org/wiki/NaN#Comparison_with_NaN
+  if (std::isnan(current) || std::isnan(next)) {
+    *increasing = false;
+    *strictly_increasing = false;
+    *decreasing = false;
+    *strictly_decreasing = false;
+  } else {
+    bool equal =
+        // Approximately equal within some error bound (epsilon).
+        (options.floating_approximate &&
+         (fabs(current - next) <=
+          static_cast<typename DataType::c_type>(options.epsilon))) ||
+        // Or exactly equal.
+        current == next;
+    if (*increasing) {
+      if (!(equal || next > current)) {
+        *increasing = false;
+        *strictly_increasing = false;
+      }
+    }
+    if (*decreasing) {
+      if (!(equal || next < current)) {
+        *decreasing = false;
+        *strictly_decreasing = false;
+      }
+    }
+    if (*strictly_increasing) {
+      if (equal || !(next > current)) {
+        *strictly_increasing = false;
+      }
+    }
+    if (*strictly_decreasing) {
+      if (equal || !(next < current)) {
+        *strictly_decreasing = false;
+      }
+    }
+  }
+}
+
+template <typename DataType>
+enable_if_not_floating_point<DataType> IsMonotonicCheck(
+    const typename DataType::c_type& current, const typename DataType::c_type& next,
```

Comment:
After reviewing this PR, I think implementations of `IsMonotonicCheck` can be categorized as follows based on `DataType`:
* have `c_type` (e.g., primitive numeric, datetime, timestamp)
   * `c_type` is floating-point
   * `c_type` is not floating-point
* do not have `c_type` (binary, string, intervals)
   * custom implementation for `binary` and `string`
   * custom implementation for intervals

There would be at least 4 type-specific implementations of `IsMonotonicCheck`, where the
the `enable_if` for this case would be of the form (pseudocode) [`enable_if_has_c_type and !not_floating_point`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/type_traits.h#L748)